init.go

```diff
@@ -200,3 +209,29 @@ func createDirectories(directories []string) error {
 	}
 	return nil
 }
+
+func shutdownPlugin() {
+	log.Info("Shutting down apidAnalytics plugin")
+
+	// Close all open files and move directories in tmp to staging.
+	bucketMaplock.RLock()
+	defer bucketMaplock.RUnlock()
+	for _, bucket := range bucketMap {
+		log.Infof("closing bucket '%s' as a part of shutdown", bucket.DirName)
+		closeGzipFile(bucket.FileWriter)
+
+		dirToBeClosed := filepath.Join(localAnalyticsTempDir, bucket.DirName)
+		stagingPath := filepath.Join(localAnalyticsStagingDir, bucket.DirName)
+		// close files in tmp folder and move directory to
+		// staging to indicate its ready for upload
+		err := os.Rename(dirToBeClosed, stagingPath)
+		if err != nil {
+			log.Errorf("Cannot move directory '%s' from"+
+				" tmp to staging folder due to '%s", bucket.DirName, err)
+		}
+
+		bucketMaplock.Lock()
```

Comment:
It's a deadlock if you `Lock()` before you  release reader lock `RUnlock ()`. 
Run this:
https://play.golang.org/p/sU5nJENJlm

Since you're shutting down, I would suggest not delete key one-by-one. Instead you can just do this at the end of this func, after the `for` loop:
```
bucketMaplock.Lock()
bucketMap = nil
bucketMaplock.Unlock()
```
Be sure you release reader lock `RUnlock()` before you you lock writer lock `Lock()`.